### PR TITLE
Secret store (part1 - key management)

### PR DESCRIPTION
**Pull request by NikVolf**
- creating keys
- loading keys (currenty only v3)
- saving keys (v3)
- managing, retrieving and caching keys withing the specified directory

encrypting/decrypting with this keys should probably be another `mod`

### CODE REVIEW

File: ethcore/src/keys_directory.rs [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+	HMacSha256
+}
+
+#[derive(Clone)]
+#[allow(non_snake_case)]
+/// Kdf of type `Pbkdf2`
+/// https://en.wikipedia.org/wiki/PBKDF2
+pub struct KdfPbkdf2Params {
+	/// desired length of the derived key, in octets
+	pub dkLen: u32,
```
Comment (arkpar) [inline ethcore/src/keys_directory.rs]: why not `dk_len`?

Comment (NikVolf) [inline ethcore/src/keys_directory.rs]: it's called `dkLen` in wikipedia/rfc

**Comment by arkpar** [comment]
This should be in util or a separate crate, not in ethcore

File: ethcore/src/keys_directory.rs (hunk 1 above)
Comment (gavofyork) [inline ethcore/src/keys_directory.rs]: it's also called `dklen` in the same wikipedia article. see the equation that reads:
```
DK = T1 || T2 || ... || Tdklen/hlen
```

i think in any case rust language conventions should take precedence.

File: ethcore/src/keys_directory.rs [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+		})
+	}
+
+	fn to_json(&self) -> Json {
+		let mut map = BTreeMap::new();
+		map.insert("dklen".to_owned(), json_from_u32(self.dkLen));
+		map.insert("salt".to_owned(), Json::String(format!("{:?}", self.salt)));
+		map.insert("prf".to_owned(), Json::String("hmac-sha256".to_owned()));
+		map.insert("c".to_owned(), json_from_u32(self.c));
+
```
Comment (gavofyork) [inline ethcore/src/keys_directory.rs]: unneeded newline

File: ethcore/src/keys_directory.rs [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+	}
+
+	fn to_json(&self) -> Json {
+		let mut map = BTreeMap::new();
+		map.insert("dklen".to_owned(), json_from_u32(self.dkLen));
+		map.insert("salt".to_owned(), Json::String(format!("{:?}", self.salt)));
+		map.insert("p".to_owned(), json_from_u32(self.p));
+		map.insert("n".to_owned(), json_from_u32(self.n));
+		map.insert("r".to_owned(), json_from_u32(self.r));
+
```
Comment (gavofyork) [inline ethcore/src/keys_directory.rs]: superfluous newline

**Comment by gavofyork** [comment]
looking good (though needed a bit more care with the docs which i corrected). `get` is currently a mutable function, presumably because the cache is a normal member and not `Cell`ed. May be worth considering giving the cache field-level mutability.

File: ethcore/src/keys_directory.rs (hunk 1 above)
Comment (NikVolf) [inline ethcore/src/keys_directory.rs]: ok, will fix it

**Comment by NikVolf** [comment]
- @gavofyork yep i will look into per-field mutability, maybe even will have to when we will be adding secret-store singleton

**Comment by NikVolf** [comment]
(all other issues addressed)

**Pull request merged**